Updates export script and links official steps for external metastore

kevxmsft opened the pull request:
Fixes errors and adds support for copying Views and Constraints

kevxmsft committed:
Fixes errors in export script and adds support for copying Views and Constraints

kevxmsft committed:
Fixes link to other doc and adds notes

kevxmsft committed:
Update hive-export-import-metastore.md

combines notes, fixes indent and punctuation, adds some clarification

msft-tacox commented on hive/hive-export-import-metastore.md:
```diff
@@ -23,73 +23,98 @@ ms.author: dkakadia
 
 #### Issue:
```
Would it make sense to put all the code in a separate repository and give a link? That way it could be easily consumed by customers via git. In fact there is a git repo for the other metastore migration tool. https://github.com/hdinsight/MetastoreMigrationTools Perhaps you could refactor a little and put both in there. Your call

msft-tacox replied:
I see everything is in the hdiconfigactions blob so I suppose you could call this resolved. Happy to discuss.

kevxmsft replied:
I have no objection to either.

msft-tacox commented on hive/hive-export-import-metastore.md:
```diff
@@ -23,73 +23,98 @@ ms.author: dkakadia
+
+> [!NOTE]
+>
+> * The script supports copying of tables and partitions. For HDInsight 4.0, it additionally covers constraints, views, and materialized views. Other metadata objects, like UDFs, must be copied manually.
```
nit: it _also_ covers constraints, views ...

kevxmsft replied:
also

msft-tacox commented on hive/hive-export-import-metastore.md:
```diff
@@ -23,73 +23,98 @@ ms.author: dkakadia
+
+> [!NOTE]
+>
+> * The script supports copying of tables and partitions. For HDInsight 4.0, it additionally covers constraints, views, and materialized views. Other metadata objects, like UDFs, must be copied manually.
```
Why can't we support migration of constraints and views in 3.6?

kevxmsft replied:
'show views' supported only as of Hive 2.2.0
constraints: as of Hive 2.1.0

msft-tacox commented on hive/hive-export-import-metastore.md:
```diff
@@ -23,73 +23,98 @@ ms.author: dkakadia
+>    alter table tablename_tmp rename to tablename;
+>
+> * This procedure preserves non-ACID table locations. You can manually edit the DDL in
+`alltables.hql`, generated from the script, to reflect any location changes. *For ACID tables, a new copy of the data will be created*.
```
Style: we should do a "Note:" in front of the "For ACID tables..." sentence, on a new line

msft-tacox commented on hive/hive-export-import-metastore.md:
```diff
@@ -23,73 +23,98 @@ ms.author: dkakadia
+>
+> * The procedure assumes that after completion, the old cluster will **not** be used any longer.
+
+#### Resolution Steps:
```
Resolution steps? Maybe Execution Steps? Resolution Steps implies we are addressing some kind of problem

kevxmsft replied:
Very strange, indeed. New section title will be "Migrate from internal metastore".

msft-tacox commented on hive/hive-export-import-metastore.md:
```diff
@@ -23,73 +23,98 @@ ms.author: dkakadia
-#### Further Reading:
-
-1) [Connect to HDInsight Cluster using SSH](https://docs.microsoft.com/en-us/azure/hdinsight/hdinsight-hadoop-linux-use-ssh-unix)
+Need to export Hive metastore and import it on another HDInsight cluster.
```
Can we make the language here a bit more customer-facing?
"Scenario: Export Hive Metastore and import it..."

msft-tacox replied:
In general, try to read the doc out loud as if you are an external customer. The language in the SQL docs are a great example like https://docs.microsoft.com/en-us/sql/relational-databases/security/permissions-database-engine?view=sql-server-ver15

kevxmsft replied:
Next update will read less like robot and more like human.

kevxmsft replied:
beep boop bop

msft-tacox commented on hive/hive-export-import-metastore.md:
````diff
@@ -23,73 +23,98 @@ ms.author: dkakadia
+> * All managed tables will become transactional if the output HDInsight version is 4.0. Optionally, make the table non-transactional by exporting the data to an external table with the property 'external.table.purge'='true'. For example,
+>
+>    ```SQL
+>    create table tablename_backup like tablename;
````
nit: the syntax highlighting is messed up here in the viewer

kevxmsft replied:
I don't see it. Do we mean the nested SQL box within the NOTE box? Is there a better style for this?

kevxmsft replied:
If you mean the stretched out rendering in VSCode, it renders better in the actual docs site.

msft-tacox commented on hive/hive-export-import-metastore.md:
```diff
@@ -23,73 +23,98 @@ ms.author: dkakadia
+
+* Import works only if the destination cluster shares the same Storage Account as the source cluster.
+
+> [!NOTE]
```
There is a lot of info in this overall `note`. Can we break it down into multiple sections or try to condense?

kevxmsft replied:
Moving a lot of this stuff to an introduction.

msft-tacox commented on hive/hive-export-import-metastore.md:
```diff
@@ -23,73 +23,98 @@ ms.author: dkakadia
+
+1) [Connect to the HDInsight Cluster using SSH](https://docs.microsoft.com/en-us/azure/hdinsight/hdinsight-hadoop-linux-use-ssh-unix), to the primary headnode.
+
+1) Download the export script from an ssh session to the cluster:
```
Lets make sure that customers only have to run wget/curl exactly once. We should have customers download and run an overall script that will download everything else on their behalf, including helper scripts or other dependencies.

msft-tacox replied:
In fact in the bigger picture, it would be ideal for the customer to run some pattern like this:
1) wget the script
2) launch exaclty 1 command

3) copy alltables.hql (can we rename this to something more explanatory? Perhaps we should actually have a wrapper script around alltables.hql that does all the setup automatically) to import-destination cluster
4) run alltables.hql or equivalent command exactly once

kevxmsft replied:
Fewer steps is a noble vision. Will attach a new wrapper script to blob path to do it all. Will have a separate "Prerequisites" section for setting up the cluster and setting Hive configuration.

msft-tacox commented on hive/hive-export-import-metastore.md:
````diff
@@ -23,73 +23,98 @@ ms.author: dkakadia
+
+        ```bash
+        hn0=$(grep hn0- /etc/hosts | xargs | cut -d' ' -f4)
+        beeline -u "jdbc:hive2://$hn0:10001/default;principal=hive/_HOST@$DOMAIN_UPPER;auth-kerberos;transportMode=http" -n "$USER@$DOMAIN" -f alltables.hql
````
Both ESP and non ESP beeline can connect using the Zookeeper quorum ("HIVE JDBC URL" in Ambari). This might make setting up the connection string easier for both export and import

kevxmsft replied:
Using the ZK string is better for high availability. But it might be more cumbersome to look up a config from Ambari. Also, this would not do better in terms of 1-step-for-all-cases, because ESP would require a USER-DOMAIN pair to kinit. 

What would be nice, though, is if we "footnote" the option for ZK quorum URL, or link the HDInsight beeline connection doc (and expect that doc to cover it).

kevxmsft replied:
Added the beeline doc as further reading material.

msft-tacox commented on hive/hive-export-import-metastore.md:
````diff
@@ -23,73 +23,98 @@ ms.author: dkakadia
+        ```
+
+        ```bash
+        hn0=$(grep hn0- /etc/hosts | xargs | cut -d' ' -f4)
````
Why can't the scripts figure out hostnames, setup connection strings, accept a user/domain for kinit, and change ambari configs?

kevxmsft replied:
They can, and they will with next update.

msft-tacox commented on hive/hive-export-import-metastore.md:
````diff
@@ -23,73 +23,98 @@ ms.author: dkakadia
+
+        ```bash
+        wget "https://hdiconfigactions.blob.core.windows.net/hivemetastoreschemaupgrade/constraints2altertable.py"
+        chmod 755 constraints2altertable.py
````
If we downloaded a helper script, it could do the chmod on the customer's behalf too

msft-tacox commented on hive/hive-export-import-metastore.md:
````diff
@@ -23,73 +23,98 @@ ms.author: dkakadia
+        ```bash
+        hn0=$(grep hn0- /etc/hosts | xargs | cut -d' ' -f4)
+        beeline -u "jdbc:hive2://$hn0:10001/default;principal=hive/_HOST@$DOMAIN_UPPER;auth-kerberos;transportMode=http" -n "$USER@$DOMAIN" -f alltables.hql
+        ```
````
Most of our docs have a short conclusion with a further reading/see also section at the end.
Generally speaking these docs should be framed more like articles so they are an easier read

kevxmsft committed:
Update hive-export-import-metastore.md

Addresses feedback from collaborators.
Primary changes are
1. Make it read more like human and less like robot TSG style.
2. Reduce required steps.

kevxmsft commented:
#sign-off
